src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.SkeletonReferenceCache.cs

```diff
@@ -351,14 +327,43 @@ private bool TryGetExisting_NoLock(MetadataReferenceProperties properties, out M
 
                     // We don't deterministically release the resulting metadata since we don't know 
                     // when we should. So we leave it up to the GC to collect it and release all the associated resources.
-                    metadata = AssemblyMetadata.CreateFromStream(stream);
+                    return AssemblyMetadata.CreateFromStream(stream, leaveOpen: false);
                 }
+            }
 
-                return metadata.GetReference(
-                    documentation: documentationProvider,
-                    aliases: aliases,
-                    embedInteropTypes: embedInteropTypes,
-                    display: _assemblyName);
+            public MetadataReference? TryGetAlreadyBuiltMetadataReference(MetadataReferenceProperties properties)
+            {
+                // lookup first and eagerly return cached value if we have it.
+                lock (_metadataReferences)
+                    return _metadataReferences.TryGetValue(properties, out var lazy) && lazy.TryGetValue(out var result) ? result : null;
+            }
+
+            public Task<MetadataReference?> GetMetadataReferenceAsync(MetadataReferenceProperties properties, CancellationToken cancellationToken)
+            {
+                AsyncLazy<MetadataReference?>? lazy;
+                lock (_metadataReferences)
+                {
+                    if (!_metadataReferences.TryGetValue(properties, out lazy))
+                    {
+                        lazy = new AsyncLazy<MetadataReference?>(c => ComputeReferenceAsync(properties, c), cacheResult: true);
```

Comment:
⚠️ This is going to cause `properties` to be captured during every call, even when the value already exists

Reply:
this code is now removed.